Source/WebCore/html/parser/HTMLDocumentParser.cpp

```diff
@@ -259,8 +259,13 @@ bool HTMLDocumentParser::pumpTokenizerLoop(SynchronousMode mode, bool parsingFra
 {
     do {
         if (UNLIKELY(isWaitingForScripts())) {
-            if (mode == AllowYield && m_parserScheduler->shouldYieldBeforeExecutingScript(m_treeBuilder->scriptToProcess(), session))
-                return true;
+            if (mode == AllowYield && m_parserScheduler->shouldYieldBeforeExecutingScript(m_treeBuilder->scriptToProcess(), session)) {
+                // Only yield if there has been progress since last yield.
+                if (session.processedTokens > session.processedTokensOnLastYield) {
+                    session.processedTokensOnLastYield = session.processedTokens;
+                    return true;
+                }
```

Comment:
Can you think of any case where this check would be false, and we would want to yield anyway? If not, then I think this check belongs inside shouldYieldBeforeExecutingScript. Because our observation is that you should not yield if you haven't made any progress yet.

Reply:
Also session.processedTokensOnLastYield is not a great name since this is only updated on a before-script yield yet there is another yield path (`shouldYieldBeforeToken()`).

Reply:
> Can you think of any case where this check would be false, and we would want to yield anyway? If not, then I think this check belongs inside shouldYieldBeforeExecutingScript. Because our observation is that you should not yield if you haven't made any progress yet.

That is a good point. I cannot think of such a case, and moved the check inside shouldYieldBeforeExecutingScript.

Thanks for reviewing!

Reply:
> Also session.processedTokensOnLastYield is not a great name since this is only updated on a before-script yield yet there is another yield path too (`shouldYieldBeforeToken()`).

Ah, yes, that is a good point. In the latest patch I have updated the name, which is hopefully more accurate.

Thanks for reviewing!